File: src/main/java/com/sy/coladay/init/CreateDatabaseTask.java
package com.sy.coladay.init;

import static java.lang.String.format;
import static java.lang.System.getenv;
import static org.apache.commons.lang3.StringUtils.isAlpha;

import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.datasource.DriverManagerDataSource;

@Slf4j
public final class CreateDatabaseTask implements Task {


  private final String dbAdminUrl;

  private final String dbAdminUser;

  private final String dbAdminPassword;

  private final String databaseName;


  public CreateDatabaseTask(@NonNull String dbAdminUrl,
                            @NonNull String dbAdminUser,
                            @NonNull String dbAdminPassword,
                            @NonNull String databaseName) {

    if (!isAlpha(databaseName)) {
      throw new IllegalArgumentException(format("Database name %s must contain letters "
                                                    + "only", databaseName));
    }

    this.dbAdminUrl = dbAdminUrl;
    this.dbAdminUser = dbAdminUser;
    this.dbAdminPassword = dbAdminPassword;
    this.databaseName = databaseName;
  }

  public CreateDatabaseTask() {

    this(getenv("DB_ADMIN_URL"),
         getenv("DB_ADMIN_USER"),
         getenv("DB_ADMIN_PASSWORD"),
         getenv("COLADAY_DB_NAME")
    );
  }

  public void exec() {

    LOG.info("Running database creation task with db admin url {}, user {} and cola db name {} ",
             this.dbAdminUrl, this.dbAdminUser, this.databaseName
    );
    try {
      var dataSource = new DriverManagerDataSource();
      dataSource.setDriverClassName("org.postgresql.Driver");
      dataSource.setUrl(dbAdminUrl);
      dataSource.setUsername(dbAdminUser);
      dataSource.setPassword(dbAdminPassword);
      var jdbcTemplate = new JdbcTemplate(dataSource);
      var dbExists = "SELECT count(*) FROM pg_database WHERE datname = ?";
      var count = jdbcTemplate.queryForObject(dbExists,
                                              new Object[]{databaseName},
                                              Integer.class
      );

      // https://www.baeldung.com/sql-injection
      // Could not use prepared statement here as this approach only works for placeholders used
      // as values. The main reason behind this is the very nature of a prepared statement:
      // database server use them to cache the query plan required to pull the result set, which
      // usually is the same for any possible value. This is not true for table names and other
      // constructs available in the SQL language such as columns used in an order by clause.
      if (count == 0) {
        jdbcTemplate.execute(format("CREATE DATABASE %s", databaseName));
        LOG.info("Database {} created", databaseName);
      } else {
        LOG.info("Skip creating database {} as it already exist", databaseName);
      }
    } catch (Throwable throwable) {
      throw new DatabaseCreationFailure(throwable);
    }
  }

  private class DatabaseCreationFailure extends RuntimeException {

    DatabaseCreationFailure(Throwable throwable) {
      super(throwable);
    }
  }
}

File: src/main/kotlin/com/sy/coladay/init/CreateDatabaseTask.kt
package com.sy.coladay.init

import org.apache.commons.lang3.StringUtils.isAlpha
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.datasource.DriverManagerDataSource
import java.lang.String.format

class CreateDatabaseTask @JvmOverloads constructor(
    dbAdminUrl: String = System.getenv("DB_ADMIN_URL"),
    dbAdminUser: String =
        System.getenv("DB_ADMIN_USER"),
    dbAdminPassword: String =
        System.getenv("DB_ADMIN_PASSWORD"),
    databaseName: String =
        System.getenv("COLADAY_DB_NAME")
) : Task {
    private val dbAdminUrl: String
    private val dbAdminUser: String
    private val dbAdminPassword: String
    private val databaseName: String

    init {
        require(isAlpha(databaseName)) {
            format("Database name %s must contain letters only", databaseName)
        }
        this.dbAdminUrl = dbAdminUrl
        this.dbAdminUser = dbAdminUser
        this.dbAdminPassword = dbAdminPassword
        this.databaseName = databaseName
    }

    override fun exec() {
        LOG.info(
            "Running database creation task with db admin url {}, user {} and cola db name {} ",
            dbAdminUrl, dbAdminUser, databaseName)

        val dataSource = DriverManagerDataSource()
        dataSource.apply {
            setDriverClassName("org.postgresql.Driver")
            url = dbAdminUrl
            username = dbAdminUser
            password = dbAdminPassword
        }

        try {
            // https://www.baeldung.com/sql-injection
            // Could not use prepared statement here as this approach only works for placeholders used
            // as values. The main reason behind this is the very nature of a prepared statement:
            // database server use them to cache the query plan required to pull the result set, which
            // usually is the same for any possible value. This is not true for table names and other
            // constructs available in the SQL language such as columns used in an order by clause.
            with(JdbcTemplate(dataSource)) {
                val dbExists = "SELECT count(*) FROM pg_database WHERE datname = ?"
                queryForObject(dbExists, arrayOf<Any>(databaseName), Int::class.java)
                    .apply {
                        when(this) {
                            0 -> {
                                execute(String.format("CREATE DATABASE %s", databaseName))
                                LOG.info("Database {} created", databaseName)
                            }
                            else -> LOG.info("Skip creating database {} as it already exist", databaseName)
                        }
                    }
            }
        } catch (throwable: Throwable) {
            throw DatabaseCreationFailure(throwable)
        }
    }

    companion object {

        private val LOG: Logger = LoggerFactory.getLogger(CreateDatabaseTask::class.java)
    }

    private inner class DatabaseCreationFailure internal constructor(throwable: Throwable?) :
        RuntimeException(throwable)
}